File: Python/App/weather/methods/weather_app.py
import requests as req
import json
import config

LIMIT = '15'


def get_json_by_city_name(city_name):
    _user_request = req.request('GET', 'http://api.openweathermap.org/geo/1.0/direct?'
                                       'q=' + city_name +
                                       '&limit=' + LIMIT +
                                       '&appid=' + config.TOKEN
                                ).text

    json_object = json.loads(_user_request)

    if len(json_object) != 0:
        print('Пожалуйста, конкретизируйте ваш выбор: ')
        element_number = 1
        for element in json_object:
            print(f'{element_number}: {element["name"]} - {element["country"]} - {element["state"]} ({element["lat"]} : {element["lon"]})')
            element_number += 1

        return json_object[get_user_choice(len(json_object)) - 1]
    else:
        raise TypeError('Неизвестный город')


def get_weather_by_coordinates(lat, lon):
    _user_request = req.request('GET', 'https://api.openweathermap.org/data/2.5/weather?'
                                       'lat=' + str(lat) +
                                       '&lon=' + str(lon) +
                                       '&appid=' + config.TOKEN +
                                       '&units=metric'
                                ).text

    main_param = json.loads(_user_request)
    temp_param = json.loads(str(main_param["main"]).replace('\'', '"'))
    print(f'Тепмература: {str(temp_param["temp"])}C°. Ощущается как: {str(temp_param["feels_like"])}C°.')


def get_user_choice(json_size):
    try:
        _user_choice = int(input('\nВыбор: '))
        if 0 < _user_choice <= json_size:
            return _user_choice
        else:
            print('К сожалению, я не понял ваш выбор или была допущена ошибка. Пожалуйста, повторите снова')
            return get_user_choice(json_size)
    except ValueError as ve:
        print('К сожалению, я не понял ваш выбор или была допущена ошибка. Пожалуйста, повторите снова')
        return get_user_choice(json_size)


def start_weather_app():
    try:
        _user_city_name = input('Введите название города: ')
        return get_json_by_city_name(_user_city_name)
    except TypeError as te:
        print(te)
        return start_weather_app()


if __name__ == '__main__':
    current_city_json = start_weather_app()
    _lat = current_city_json["lat"]
    _lon = current_city_json["lon"]
    get_weather_by_coordinates(_lat, _lon)

File: Java/EducationProject/src/main/java/app/WeatherApp.java
package app;

import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import java.io.IOException;
import java.net.URL;
import java.text.SimpleDateFormat;
import java.util.Map;


public class WeatherApp {
    private final String weatherApiToken;
    private static final int LIMIT = 15;

    public WeatherApp(String weatherApiToken) {
        this.weatherApiToken = weatherApiToken;
    }

    public Map<Long, String> getWeatherByCoordinates(String lat, String lon) throws IOException {
        OkHttpClient client = new OkHttpClient();
        Request request = new Request.Builder()
                .url(new URL("https://api.openweathermap.org/data/2.5/weather?" +
                        "lat=" + lat +
                        "&lon=" +  lon +
                        "&appid=" + weatherApiToken +
                        "&units=metric"))
                .build();

        Response response = client.newCall(request).execute();
        String result = response.body().string();

        Gson gson = new Gson();
        JsonObject mainJson = gson.fromJson(result, JsonObject.class);
        JsonObject tempJson = gson.fromJson(mainJson.get("main"), JsonObject.class);
        return Map.of(mainJson.get("id").getAsLong(), String.format("Температура: %sC°. Ощущается как: %sC°\n", tempJson.get("temp"), tempJson.get("feels_like")));
    }

    public JsonArray getJsonByCityName(String cityName) throws IOException {
        OkHttpClient client = new OkHttpClient();
        Request request = new Request.Builder()
                .url(new URL("http://api.openweathermap.org/geo/1.0/direct?" +
                        "q=" + cityName +
                        "&limit=" + LIMIT +
                        "&appid=" + weatherApiToken))
                .build();

        Response response = client.newCall(request).execute();
        String result = response.body().string();

        Gson gson = new Gson();
        return gson.fromJson(result, JsonArray.class);
    }

    public String getCitiesList(JsonArray mainJson) {
        StringBuilder resultMessage = new StringBuilder();
        if (mainJson.size() != 0) {
            resultMessage.append("Пожалуйста, конкретизируйте ваш выбор: \n");
            int elementNumber = 1;
            for (JsonElement element : mainJson) {
                resultMessage.append(String.format("%d: %s - %s - %s (%s : %s)\n",
                        elementNumber,
                        element.getAsJsonObject().get("name").getAsString(),
                        element.getAsJsonObject().get("country").getAsString(),
                        element.getAsJsonObject().get("state").getAsString(),
                        element.getAsJsonObject().get("lat").getAsString(),
                        element.getAsJsonObject().get("lon").getAsString()
                ));
                elementNumber++;
            }
            return resultMessage.toString();
        } else {
            throw new IllegalArgumentException("Неизвестный город");
        }
    }
}
